src/object/cli_obj.c

```diff
@@ -984,7 +984,8 @@ obj_shard_tgts_query(struct dc_object *obj, uint32_t map_ver, uint32_t shard,
 	shard_tgt->st_shard	= shard;
 	shard_tgt->st_shard_id	= obj_shard->do_id.id_shard;
 	shard_tgt->st_tgt_idx	= obj_shard->do_target_idx;
-	if (obj_auxi->cond_modify && (obj_shard->do_rebuilding || obj_shard->do_reintegrating))
+	if ((obj_auxi->cond_modify || obj_is_punch_opc(obj_auxi->opc)) &&
```

Comment:
probably only punch with DAOS_COND_PUNCH need to delay forward? so seems original check "if (obj_auxi->cond_modify && (obj_shard->do_rebuilding || obj_shard->do_reintegrating))" is good enough and need not change here?